### ISSUE TITLE
Spin CLI: Fingerprint resources to protect against concurrent writes

### ISSUE THREAD

**Issue by jtk54** [spinnaker/spinnaker#2966]

**Comment by spinnakerbot** [comment]
This issue hasn't been updated in 45 days, so we are tagging it as 'stale'. If you want to remove this label, comment:

> @spinnakerbot remove-label stale

**Comment by spinnakerbot** [comment]
This issue hasn't been updated in 52 days, so we are tagging it as 'stale'. If you want to remove this label, comment:

> @spinnakerbot remove-label stale

**Comment by spinnakerbot** [comment]
This issue hasn't been updated in 45 days, so we are tagging it as 'stale'. If you want to remove this label, comment:

> @spinnakerbot remove-label stale

**Comment by spinnakerbot** [comment]
This issue hasn't been updated in 45 days, so we are tagging it as 'stale'. If you want to remove this label, comment:

> @spinnakerbot remove-label stale

**Comment by jtk54** [comment]
@spinnakerbot no-lifecycle

**Comment by spinnakerbot** [comment]
This issue hasn't been updated in 45 days, so we are tagging it as 'stale'. If you want to remove this label, comment:

> @spinnakerbot remove-label stale

**Comment by jtk54** [comment]
@spinnakerbot no-lifecycle

**Comment by patpatpat123** [comment]
# #2966: Fingerprinting resources to protect against concurrent writes

Hi folks 👋

I went looking into this one because it's been open a while and looked like a small
CLI gap. It turned out to be a bit more interesting than that, so I wanted to write up
what I found along with a concrete patch, rather than just drop a PR on you.

**Short version:** we can't fix this in `spin` alone. The server-side mechanism `spin`
would need — front50's `?staleCheck=true` — has been silently inert since August 2022.
It doesn't fire for *any* HTTP client, including Deck, which has been sending
`staleCheck: true` on every pipeline save for about five years with no effect.

So the good news is the fix is small. The less-good news is it's in front50, not the CLI.

---

## Contents

- [What's actually broken](#whats-actually-broken)
- [When it broke](#when-it-broke)
- [Why a `spin`-only fix would be a placebo](#why-a-spin-only-fix-would-be-a-placebo)
- [The plan (3 PRs)](#the-plan-3-prs)
- [PR 1 — Unblock the build](#pr-1--unblock-the-build)
- [PR 2 — Repair `staleCheck` in front50](#pr-2--repair-stalecheck-in-front50)
- [PR 3 — Opt `spin` in](#pr-3--opt-spin-in)
- [Tests](#tests)
- [Things I could not verify](#things-i-could-not-verify)
- [Known limitations, stated up front](#known-limitations-stated-up-front)
- [Questions for maintainers](#questions-for-maintainers)

---

## What's actually broken

front50 already has optimistic concurrency control for pipelines. `PipelineController`
takes a `staleCheck` query param, and when it's on it compares the `lastModified` you
submitted against the one currently stored:

```java
// front50/front50-web/src/main/java/com/netflix/spinnaker/front50/controllers/PipelineController.java:547-552
// Run stale pipeline definition check
if (staleCheck
    && !Strings.isNullOrEmpty(pipeline.getId())
    && pipeline.getLastModified() != null) {
  checkForStalePipeline(pipeline, errors);
}
```

That third condition is the problem. **`pipeline.getLastModified()` can never be
non-null for an HTTP-submitted pipeline**, so `checkForStalePipeline` is never called.

Here's the chain:

**1. `lastModified` can't leave the server.** It's `@JsonIgnore`d on the very
ObjectMapper Spring MVC uses:

```java
// front50/front50-core/src/main/java/com/netflix/spinnaker/front50/jackson/mixins/PipelineMixins.java:85
@JsonIgnore private String lastModified;
```

(registered at `Front50ApiModule.java:34` via `setMixInAnnotations(Pipeline.class, PipelineMixins.class)`)

**2. A substitute does leave — but can't come back.** `Pipeline` has an `updateTs`
property that *is* serialized, and it's derived from `lastModified`:

```java
// front50/front50-api/src/main/java/com/netflix/spinnaker/front50/api/model/pipeline/Pipeline.java:65-69
public String getUpdateTs() {
  var lastModified = getLastModified();

  return lastModified != null ? lastModified.toString() : null;
}
```

So clients *do* receive a perfectly good fingerprint. But there is **no `setUpdateTs`**,
and the backing field at line 45 has no `@Setter` (unlike every field around it):

```java
private String createTs;
private String updateTs;      // <- no @Setter, and no hand-written setter either
private String lastModifiedBy;
@Getter @Setter private Long lastModified;
```

Jackson sees a getter-only property and quietly drops `updateTs` on the way in. The
round trip is open at one end.

**3. There's a second break behind the first.** `checkForStalePipeline` reads the stored
side through `pipelineDAO.findById` → `StorageServiceSupport.findById` →
`service.loadObject(...)`. And the single-object SQL read doesn't select
`last_modified_at`:

```kotlin
// front50/front50-sql/src/main/kotlin/com/netflix/spinnaker/front50/model/SqlStorageService.kt:102-126
override fun <T : Timestamped> loadObject(objectType: ObjectType, objectKey: String): T {
  val result = withPool(poolName) {
    jooq.withRetry(sqlRetryProperties.reads) { ctx ->
      ctx
        .select(
          field("body", String::class.java),
          field("created_at", Long::class.java)     // <- no last_modified_at
        )
        ...
  return objectMapper.readValue(...).apply {
    this.createdAt = result.get(field("created_at", Long::class.java))
                                                     // <- lastModified never assigned
  }
}
```

`loadObjects` (bulk, line 158) and `listObjectVersions` (line 429) both *do* assign it —
`loadObject` is the outlier. So if you only fixed the submitted side, you'd get
`submitted.equals(null)` → `false` → **every save rejected**. Both halves need fixing
together.

## When it broke

`git log` points at front50 [#1142](https://github.com/spinnaker/front50/pull/1142)
(`5c15d0c9ba`, 2022-08-02, *"fix: Revision history is not showing the timestamp of the
revision"*). One of its commits is literally *"fix: replace updateTs with lastModified"*.
[#1159](https://github.com/spinnaker/front50/pull/1159) then removed the `@JsonIgnore`
from `updateTs`, leaving today's dead field.

No blame intended — this is exactly the kind of thing that's invisible without a test
that goes through the HTTP layer, and the existing test doesn't
(`PipelineControllerSpec.groovy` builds the controller with a plain `new ObjectMapper()`
and calls methods directly, so it never sees the mixin). That's worth fixing too, and
PR 2 does.

## Why a `spin`-only fix would be a placebo

Today `spin pipeline save` GETs the pipeline purely to steal its `id`, then blindly
overwrites:

```go
// spin/cmd/pipeline/save.go:88-102
foundPipeline, queryResp, _ := options.GateClient.ApplicationControllerAPI.GetPipelineConfig(...).Execute()
switch queryResp.StatusCode {
case http.StatusOK:
    if len(foundPipeline) > 0 {
        pipelineJson["id"] = foundPipeline["id"].(string)   // only the id
    }
...
saveResp, err := options.GateClient.PipelineControllerAPI.SavePipeline(...).RequestBody(pipelineJson).Execute()
```

If we just taught it to send `staleCheck=true`, the CLI would advertise concurrency
protection, send the parameter, and clobber anyway. I'd rather not ship that — silent
false confidence is worse than the honest gap we have now.

## The plan (3 PRs)

| | What | Why separate |
|---|---|---|
| **PR 1** | Gradle 9 `exec {}` migration | Pure build fix, unrelated to this issue, currently blocking all `spin` work |
| **PR 2** | Repair `staleCheck` in front50 | Stands alone — fixes Deck too. Testable in CI today. |
| **PR 3** | Opt `spin` in | Small once PR 2 lands |

Happy to combine or split differently — just tell me what you'd prefer to review.

---

## PR 1 — Unblock the build

`spin` doesn't build right now. The wrapper pins Gradle 9.6.1, which removed
`Project.exec()`. Four consecutive `spin.yml` runs have failed since 2026-08-01
(last green 2026-07-16, four days before the Gradle 9 upgrade in `b58c1be397`):

```
> Could not find method exec() for arguments [...] on task ':gate:swagger'
```

`keel-sql` was already migrated and has the pattern we should copy
(`keel/keel-sql/keel-sql.gradle:14-18`). Nine call sites remain.

This must be **one PR touching both files** — `.github/workflows/spin.yml` is
path-filtered to `spin/**`, so a gate-only change would never be validated by the
spin build.

### `gate/build.gradle`

```diff
--- a/gate/build.gradle
+++ b/gate/build.gradle
@@ -1,3 +1,6 @@
+import javax.inject.Inject
+import org.gradle.process.ExecOperations
+
 plugins {
   id 'io.spinnaker.project' apply false
   alias(libs.plugins.kotlin.jvm) apply false
@@ -66,11 +69,17 @@
   }
 }
 
+// Gradle 9 removed Project.exec(). Use the injected ExecOperations service instead.
+abstract class ExecOperationsHolder {
+  @Inject abstract ExecOperations getExecOps()
+}
+def execOps = objects.newInstance(ExecOperationsHolder).execOps
+
 // This is registered here so that 'gate' is the default working directory for the commands in this script
 tasks.register('swagger') {
   doLast {
-    exec {
+    execOps.exec {
       workingDir project.rootDir
       commandLine('bash', '-c', 'swagger/generate_swagger.sh')
     }
   }
 }
```

### `spin/build.gradle`

```diff
--- a/spin/build.gradle
+++ b/spin/build.gradle
@@ -1,3 +1,12 @@
+import javax.inject.Inject
+import org.gradle.process.ExecOperations
+
+// Gradle 9 removed Project.exec(). Use the injected ExecOperations service instead.
+abstract class ExecOperationsHolder {
+    @Inject abstract ExecOperations getExecOps()
+}
+def execOps = objects.newInstance(ExecOperationsHolder).execOps
+
 // Generate Gate swagger JSON
 tasks.register('gateswagger') {
     dependsOn gradle.parent.includedBuild('gate').task(':swagger')
@@ -10,7 +19,7 @@
     onlyIf {
         def noDockerMsg = "Could not find a Docker environment - skipping Spin CLI build"
         try {
-            def result = exec {
+            def result = execOps.exec {
                 ignoreExitValue = true
                 commandLine 'docker', '-v'
             }
@@ -25,7 +34,7 @@
     }
 
     doLast {
-        exec {
+        execOps.exec {
             workingDir project.projectDir
             commandLine 'bash', '-c', 'swagger/generate_swagger.sh'
         }
@@ -44,7 +53,7 @@
     onlyIf {
         def noGoMsg = "Could not find installed golang - skipping Spin CLI build"
         try {
-            def result = exec {
+            def result = execOps.exec {
                 ignoreExitValue = true
                 commandLine 'go', 'version'
             }
@@ -61,7 +70,7 @@
 
 tasks.register('clean') {
     doLast {
-        exec {
+        execOps.exec {
             commandLine 'rm', '-rf', './gateapi'
         }
     }
@@ -81,7 +90,7 @@
 tasks.register('godeps') {
     dependsOn 'gateapi'
     doLast {
-        exec {
+        execOps.exec {
             commandLine 'go', 'get', '-d', '-v'
         }
     }
@@ -90,7 +99,7 @@
 tasks.register('gobuild') {
     dependsOn 'godeps'
     doLast {
-        exec {
+        execOps.exec {
             commandLine 'go', 'build', '.'
         }
     }
@@ -99,7 +108,7 @@
 tasks.register('gotest') {
     dependsOn 'godeps'
     doLast {
-        exec {
+        execOps.exec {
             commandLine 'go', 'test', '-v', './...'
         }
     }
@@ -131,7 +140,7 @@
             dist/$VERSION/linux/amd64/spin --version
         """
 
-        exec {
+        execOps.exec {
             commandLine 'bash', '-c', cmd
         }
     }
```

> **Heads up while we're in here:** `.github/workflows/spin.yml` invokes
> `:spin:gorelease`, which declares only `mustRunAfter 'gotest'` — never `dependsOn`.
> So `go test ./...` has **never run in CI**. Worth fixing, but I'd suggest a raw
> `go test ./...` step rather than routing through `godeps`, whose bare `go get -d -v`
> is an untested unknown against Go 1.23. Happy to do that here or separately.

---

## PR 2 — Repair `staleCheck` in front50

Two small changes. This one fixes Deck as well as the CLI, and it's independently
valuable regardless of what we decide about `spin`.

### `Pipeline.java` — let the fingerprint come home

```diff
--- a/front50/front50-api/src/main/java/com/netflix/spinnaker/front50/api/model/pipeline/Pipeline.java
+++ b/front50/front50-api/src/main/java/com/netflix/spinnaker/front50/api/model/pipeline/Pipeline.java
@@ -65,6 +65,29 @@ public class Pipeline implements Timestamped, Cloneable {
   public String getUpdateTs() {
     var lastModified = getLastModified();
 
     return lastModified != null ? lastModified.toString() : null;
   }
 
+  /**
+   * Counterpart to {@link #getUpdateTs()}.
+   *
+   * <p>Without this, {@code updateTs} is a read-only derived property: clients receive it but
+   * cannot echo it back, so {@link #getLastModified()} is always null on an HTTP-submitted
+   * pipeline and the {@code ?staleCheck=true} optimistic-concurrency check in {@code
+   * PipelineController#validatePipeline} silently never runs.
+   *
+   * <p>Null, blank and unparseable values are treated as "no fingerprint supplied" rather than as
+   * an error, so a hand-written or hand-edited pipeline body still saves — it just isn't
+   * stale-checked. The value submitted here is only ever used for comparison; {@code
+   * StorageServiceSupport#update} re-stamps {@code lastModified} from the server clock before
+   * anything is persisted, so a client cannot forge a stored timestamp.
+   */
+  public void setUpdateTs(String updateTs) {
+    if (updateTs == null || updateTs.trim().isEmpty()) {
+      setLastModified(null);
+      return;
+    }
+    try {
+      setLastModified(Long.valueOf(updateTs.trim()));
+    } catch (NumberFormatException e) {
+      setLastModified(null);
+    }
+  }
+
   public void setAny(String key, Object value) {
     anyMap.put(key, value);
   }
```

Two deliberate choices worth flagging for review:

- **No new dependency.** `front50-api` has no Jackson or Guava on its compile classpath,
  so this is plain `Long.valueOf` in a try/catch rather than `Longs.tryParse`.
- **Please don't "clean up" the now-truly-dead `private String updateTs;` field at line
  45.** `PipelineMixins` anchors its `@JsonInclude(Include.NON_NULL)` rule to that
  member; deleting the field deletes the inclusion rule.

### `SqlStorageService.kt` — populate the stored side

```diff
--- a/front50/front50-sql/src/main/kotlin/com/netflix/spinnaker/front50/model/SqlStorageService.kt
+++ b/front50/front50-sql/src/main/kotlin/com/netflix/spinnaker/front50/model/SqlStorageService.kt
@@ -102,7 +102,8 @@ class SqlStorageService(
   override fun <T : Timestamped> loadObject(objectType: ObjectType, objectKey: String): T {
     val result = withPool(poolName) {
       jooq.withRetry(sqlRetryProperties.reads) { ctx ->
         ctx
           .select(
             field("body", String::class.java),
-            field("created_at", Long::class.java)
+            field("created_at", Long::class.java),
+            field("last_modified_at", Long::class.java)
           )
           .from(definitionsByType[objectType]!!.tableName)
           .where(
@@ -120,6 +121,7 @@ class SqlStorageService(
     return objectMapper.readValue(
       result.get(field("body", String::class.java)),
       objectType.clazz as Class<T>
     ).apply {
       this.createdAt = result.get(field("created_at", Long::class.java))
+      this.lastModified = result.get(field("last_modified_at", Long::class.java))
     }
   }
```

This just brings `loadObject` in line with `loadObjects` (:158) and
`listObjectVersions` (:429), which already do exactly this.

---

## PR 3 — Opt `spin` in

Once PR 2 lands, this is genuinely small.

### `spin/cmd/pipeline/save.go`

```diff
--- a/spin/cmd/pipeline/save.go
+++ b/spin/cmd/pipeline/save.go
@@ -15,9 +15,11 @@
 package pipeline
 
 import (
+	"errors"
 	"fmt"
 	"io/ioutil"
 	"net/http"
+	"strings"
 
 	"github.com/spf13/cobra"
 
@@ -27,6 +29,7 @@ type saveOptions struct {
 	*PipelineOptions
 	output       string
 	pipelineFile string
+	staleCheck   bool
 }
 
@@ -52,6 +55,9 @@ func NewSaveCmd(pipelineOptions *PipelineOptions) *cobra.Command {
 	cmd.PersistentFlags().StringVarP(&options.pipelineFile, "file", "f", "", "path to the pipeline file")
+	cmd.PersistentFlags().BoolVar(&options.staleCheck, "stale-check", true,
+		"fail instead of overwriting if the pipeline changed in Spinnaker after this command read it")
 
 	return cmd
 }
@@ -85,26 +91,52 @@ func savePipeline(cmd *cobra.Command, options *saveOptions) error {
 	application := pipelineJson["application"].(string)
 	pipelineName := pipelineJson["name"].(string)
 
-	foundPipeline, queryResp, _ := options.GateClient.ApplicationControllerAPI.GetPipelineConfig(options.GateClient.Context, application, pipelineName).Execute()
-	switch queryResp.StatusCode {
+	// updateTs is server-owned. A value carried in the user's input file (e.g. a committed
+	// export) must never become a precondition -- only the value we just read may be.
+	delete(pipelineJson, "updateTs")
+
+	foundPipeline, queryResp, queryErr := options.GateClient.ApplicationControllerAPI.
+		GetPipelineConfig(options.GateClient.Context, application, pipelineName).Execute()
+	if queryResp == nil {
+		return fmt.Errorf("failed to look up pipeline %q in application %q: %w",
+			pipelineName, application, queryErr)
+	}
+	defer queryResp.Body.Close()
+
+	switch queryResp.StatusCode {
 	case http.StatusOK:
 		// pipeline found, let's use Spinnaker's known Pipeline ID, otherwise we'll get one created for us
 		if len(foundPipeline) > 0 {
-			pipelineJson["id"] = foundPipeline["id"].(string)
+			id, ok := foundPipeline["id"].(string)
+			if !ok {
+				return fmt.Errorf("pipeline %q in application %q has no usable id", pipelineName, application)
+			}
+			pipelineJson["id"] = id
+
+			// Carry the server's fingerprint forward so front50 can reject the write if
+			// someone else changed the pipeline inside our read -> write window.
+			if options.staleCheck {
+				if updateTs, present := foundPipeline["updateTs"]; present && updateTs != nil {
+					pipelineJson["updateTs"] = updateTs
+				}
+			}
 		}
 	case http.StatusNotFound:
 		// pipeline doesn't exists, let's create a new one
 	default:
 		b, _ := ioutil.ReadAll(queryResp.Body)
 		return fmt.Errorf("unhandled response %d: %s", queryResp.StatusCode, b)
 	}
 
-	saveResp, err := options.GateClient.PipelineControllerAPI.SavePipeline(options.GateClient.Context).RequestBody(pipelineJson).Execute()
-	if err != nil {
-		return err
-	}
-	if saveResp.StatusCode != http.StatusOK {
-		return fmt.Errorf("Encountered an error saving pipeline, status code: %d\n", saveResp.StatusCode)
-	}
+	saveReq := options.GateClient.PipelineControllerAPI.
+		SavePipeline(options.GateClient.Context).RequestBody(pipelineJson)
+	if options.staleCheck {
+		saveReq = saveReq.StaleCheck(true)
+	}
+
+	saveResp, err := saveReq.Execute()
+	if err != nil {
+		return savePipelineError(application, pipelineName, saveResp, err)
+	}
+	if saveResp.StatusCode != http.StatusOK {
+		return savePipelineError(application, pipelineName, saveResp, nil)
+	}
 
 	options.Ui.Success("Pipeline save succeeded")
 	return nil
 }
+
+// savePipelineError turns a failed save into something actionable, special-casing
+// front50's stale-pipeline rejection.
+func savePipelineError(application, name string, resp *http.Response, err error) error {
+	body := ""
+	// The generated client drains the response body into its own error type, so prefer
+	// that when it's present. Matched structurally to keep this package free of gateapi.
+	var apiErr interface{ Body() []byte }
+	if err != nil && errors.As(err, &apiErr) {
+		body = string(apiErr.Body())
+	} else if resp != nil && resp.Body != nil {
+		if b, readErr := ioutil.ReadAll(resp.Body); readErr == nil {
+			body = string(b)
+		}
+	}
+
+	if strings.Contains(body, "is stale") {
+		return fmt.Errorf(
+			"pipeline %q in application %q was modified in Spinnaker while this command was running; nothing was saved.\n"+
+				"Re-run to pick up the change, or pass --stale-check=false to overwrite it.\n"+
+				"Server said: %s",
+			name, application, strings.TrimSpace(body))
+	}
+
+	if err != nil {
+		return fmt.Errorf("encountered an error saving pipeline %q: %w", name, err)
+	}
+	return fmt.Errorf("encountered an error saving pipeline %q, status code: %d\n%s",
+		name, resp.StatusCode, strings.TrimSpace(body))
+}
```

### `spin/cmd/pipeline/update.go`

`update` is the one true read-modify-write in the CLI — it re-POSTs the fetched map
verbatim, so it already carries `updateTs`. It just needs to ask for the check.

```diff
--- a/spin/cmd/pipeline/update.go
+++ b/spin/cmd/pipeline/update.go
@@ -24,6 +24,7 @@ type updateOptions struct {
 	*PipelineOptions
 	disabled    bool
 	enabled     bool
 	application string
 	name        string
+	staleCheck  bool
 }
@@ -56,6 +57,8 @@ func NewUpdateCmd(pipelineOptions *PipelineOptions) *cobra.Command {
 	cmd.PersistentFlags().BoolVarP(&options.disabled, "disabled", "d", false, "enable or disable pipeline")
 	cmd.PersistentFlags().BoolVarP(&options.enabled, "enabled", "e", false, "enable or disable pipeline")
+	cmd.PersistentFlags().BoolVar(&options.staleCheck, "stale-check", true,
+		"fail instead of overwriting if the pipeline changed in Spinnaker after this command read it")
 
 	return cmd
 }
@@ -63,9 +66,13 @@ func updatePipeline(cmd *cobra.Command, options *updateOptions) error {
 	application := options.application
 	pipelineName := options.name
 
-	foundPipeline, queryResp, _ := options.GateClient.ApplicationControllerAPI.GetPipelineConfig(options.GateClient.Context, application, pipelineName).Execute()
-	if queryResp.StatusCode == http.StatusNotFound {
+	foundPipeline, queryResp, queryErr := options.GateClient.ApplicationControllerAPI.
+		GetPipelineConfig(options.GateClient.Context, application, pipelineName).Execute()
+	if queryResp == nil {
+		return fmt.Errorf("failed to look up pipeline %q in application %q: %w",
+			pipelineName, application, queryErr)
+	}
+	if queryResp.StatusCode == http.StatusNotFound {
 		return fmt.Errorf("Pipeline %s not found under application %s", pipelineName, application)
 	}
@@ -85,13 +92,18 @@ func updatePipeline(cmd *cobra.Command, options *updateOptions) error {
-	saveResp, err := options.GateClient.PipelineControllerAPI.SavePipeline(options.GateClient.Context).RequestBody(foundPipeline).Execute()
-	if err != nil {
-		return err
-	}
-
-	if saveResp.StatusCode != http.StatusOK {
-		return fmt.Errorf("Encountered an error saving pipeline, status code: %d\n", saveResp.StatusCode)
-	}
+	saveReq := options.GateClient.PipelineControllerAPI.
+		SavePipeline(options.GateClient.Context).RequestBody(foundPipeline)
+	if options.staleCheck {
+		saveReq = saveReq.StaleCheck(true)
+	}
+
+	saveResp, err := saveReq.Execute()
+	if err != nil {
+		return savePipelineError(application, pipelineName, saveResp, err)
+	}
+	if saveResp.StatusCode != http.StatusOK {
+		return savePipelineError(application, pipelineName, saveResp, nil)
+	}
 
 	options.Ui.Success("Pipeline update succeeded")
 	return nil
 }
```

### What the user sees

```
$ spin pipeline save --file pipeline.json
pipeline "my-pipeline" in application "myapp" was modified in Spinnaker while this
command was running; nothing was saved.
Re-run to pick up the change, or pass --stale-check=false to overwrite it.
Server said: {"timestamp":1754870400000,"status":400,"error":"Bad Request",
"message":"The submitted pipeline is stale.  submitted updateTs 1520879791608 does
not match stored updateTs 1520879800123"}
```

**Is default-on a breaking change?** I don't think so in any normal workflow. `spin`
fetches the fingerprint itself immediately before writing, and strips any `updateTs`
from your input file, so the check can only fail when another writer commits inside
`spin`'s own read→write window — exactly the race this issue is about. Creating a new
pipeline is unaffected (front50 skips the check when `id` or `lastModified` is absent).
Against an unpatched front50 the param is ignored and behaviour is identical to today.

That said, if you'd rather ship it default-off for a release first, that's a one-word
change and I'm happy to do it.

---

## Tests

**front50 (these run in CI today, so they're the real safety net):**

```groovy
// PipelineSpec.groovy — the round trip
def 'deserializing updateTs populates lastModified'() {
  expect:
  new ObjectMapper().readValue(
    '{"id":"1","name":"sky","application":"almond","updateTs":"1520879791608"}',
    Pipeline).lastModified == 1520879791608L
}
```

```groovy
// PipelineControllerSpec.groovy — MUST go through the autowired mockMvc,
// not MockMvcBuilders.standaloneSetup and not a hand-rolled `new ObjectMapper()`.
// Only the autowired path uses the mixin-carrying converters, which is precisely
// why the existing spec didn't catch this regression.
def 'save with staleCheck rejects a pipeline carrying an outdated updateTs'()
def 'save with staleCheck accepts a pipeline carrying the current updateTs'()
def 'save without staleCheck ignores updateTs entirely'()
```

Plus a `SqlStorageService` case asserting `loadObject` and `loadObjects` agree on
`lastModified` — that pins the second half of the fix.

**spin:** `save_test.go` / `update_test.go` cases asserting `?staleCheck=true` is on the
wire and that the fetched `updateTs` survives into the POST body. The existing
`NewTestBufferHandlerFunc` (`spin/util/test_helpers.go:46`) captures the body but not the
query string, so it needs a small sibling that also records `r.URL.Query()`.

Note the existing `TestPipelineSave_json` / `_yaml` fixtures will need updating, since
`spin` now strips `updateTs` from the input file before POSTing.

---

## Things I could not verify

Being upfront about these:

1. **Whether the generated Go client exposes `.StaleCheck(bool)`.** `gate` declares the
   `@RequestParam` (`PipelineController.groovy:112-115`), so OpenAPI Generator *should*
   emit a fluent setter — but `spin/gateapi/` is generated at build time and gitignored,
   and confirming it requires `./gradlew :spin:gateapi`, which is blocked behind PR 1.
   If the name differs, it's a one-line change in PR 3.
2. **Non-SQL storage backends.** `AzureStorageService` and `OracleStorageService` also
   don't populate `lastModified` on a single-object load. The `setUpdateTs` write-through
   should cover them via the stored body, but I haven't been able to exercise them.
   Would appreciate a second opinion from anyone running those.

## Known limitations, stated up front

- **Millisecond resolution.** The token is `System.currentTimeMillis()` from the writing
  replica. Two writes in the same millisecond, or clock-skewed replicas, are
  indistinguishable. Narrow, but real — an opaque monotonic version column would be
  strictly better and is a much bigger change.
- **`optimize-cache-refreshes`.** `SqlStorageService.loadObjectsNewerThan` also never
  assigns `lastModified`, so deployments with that flag on may still see gaps. Not
  addressed here; flagging it.
- **Deck's declared type is wrong.** `deck/packages/core/src/domain/IPipeline.ts:37` says
  `updateTs?: number`, but front50 emits a JSON *string*. Harmless today because nothing
  reads it, but worth a follow-up.
- **Only pipelines.** `application save`, `project save`, `pipeline-template save` and
  `canary-config save` are all still unguarded read-modify-writes. `application save`
  doesn't even check for an existing record — there's a `// TODO(jacobkiefer)` about it
  at `spin/cmd/application/save.go:62`. `lastModified` lives on `Timestamped` in the
  published `front50-api`, so every domain object already carries a fingerprint; the gap
  is that only `PipelineController` accepts a `staleCheck` param. Happy to follow up if
  there's appetite.

## Questions for maintainers

1. Does the 3-PR split work, or would you rather see PR 1 (the Gradle fix) go in on its
   own immediately since it's unblocking all of `spin` CI?
2. `--stale-check` default-on or default-off for the first release?
3. Would you like a `controller.pipeline.save.stale-check-enabled` kill switch in
   front50? PR 2 silently activates a check that's been dormant for three years, and
   right now the only rollback is revert-and-redeploy. I left it out to keep the diff
   honest, but I can see the argument.
4. Should I open a separate issue for the front50 regression itself? It affects Deck
   independently of the CLI, and folks may want to know their pipeline saves haven't
   been stale-checked since 2022.

Thanks for reading all of this — happy to open any or all of these as PRs, or to hand
the findings off if someone closer to front50 would rather take it. 🙏

**Comment by jasonmcintosh** [comment]
JUST saw this ;) IF You don't mind - new issues help so i can catch these better!  I've got a PR to fix the spin-cli issue already ;) Though it's only a PARTIAL fix (fixes the gate:swagger hadn't got to the spin cli issue yet).  I'd welcome fixes for the rest!  I'd suggest using the "simple" approach on the gradle changes like I did on https://github.com/spinnaker/spinnaker/pull/7882/changes ;) but... that's a personal bias. Reference: https://stackoverflow.com/questions/79720795/replacing-deprecated-projectexec-in-dofirst-dolast/79720796#79720796

Stale check... I'd go default-off for the upcoming 2026.3.0 release since that's current behavior, despite being "broken" with an option to make it default, and then when 2026.3.0 hits, we adjust the flag to make it default.  Add to release notes this change.

SO plan:
* Fix build - as said already started this one, just initial work, but welcome to have some PRs for this ;) AND get it so this is better handled in CI.
* Fix front50 & issues there as a PR, with a default on the "current" behavior for the upcoming 2026.3.0 release with the allowance of a flag to re-enable this.  
* Separate PR to make that default be enabled POST go live so 2026.4.0 will get this turned on by default.

Appreciate the detailed analysis, tracking and findings and definitely looking forward to getting this fixed!!  I wouldn't mind a copy/paste to a NEW issue vs. THIS issue :p but I'm watching it now...

**Comment by jasonmcintosh** [comment]
Fixing the spin cli builds holistically now actually ;) soooo shouldn't have to worry about that one

**Comment by jasonmcintosh** [comment]
NOTE:  Would ALSO welcome fixing the go test stuff ;) I hit the same when I did a big cleanup recently but hadn't fixed adding it to run tests on CI which... I should have done and didn't get to :) 
```
Heads up while we're in here: .github/workflows/spin.yml invokes
:spin:gorelease, which declares only mustRunAfter 'gotest' — never dependsOn.
So go test ./... has never run in CI. Worth fixing, but I'd suggest a raw
go test ./... step rather than routing through godeps, whose bare go get -d -v
is an untested unknown against Go 1.23. Happy to do that here or separately.
```
AKA this ;) SO please!!!

**Comment by jasonmcintosh** [comment]
https://github.com/spinnaker/spinnaker/pull/7882/changes is merged which should fix the spin build issues.

**Comment by jasonmcintosh** [comment]
@patpatpat123 any chance you wanted to do some of these PRs? :)
